foreman/data_refinery_foreman/foreman/main.py

```diff
@@ -0,0 +1,260 @@
+    """If a job fails too many times, log it and stop retrying."""
+    # Not strictly retried but will prevent the job from getting
+    # retried any more times.
+    job.retried = True
+
+    # success may already be False, but if it was a hung or lost job
+    # this will ensure it's marked as failed.
+    job.success = False
+    job.save()
+
+    # At some point this should become more noisy/attention
+    # grabbing. However for the time just logging should be sufficient
+    # because all log messages will be closely monitored during early
+    # testing stages.
+    logger.warn("%s #%d failed %d times!!!", job.__class__.__name__, job.id, MAX_NUM_RETRIES + 1)
+
+
+def handle_downloader_jobs(jobs: List[DownloaderJob]) -> None:
+    """For each job in jobs, either retry it or log it."""
+    for job in jobs:
+        if job.num_retries < MAX_NUM_RETRIES:
+            requeue_downloader_job(job)
+        else:
+            handle_repeated_failure(job)
+
+
+def do_forever(min_loop_time: timedelta) -> Callable:
+    """Run the wrapped function in a loop forever.
+
+    The function won't be run more often than once per min_loop_time,
+    however if it takes longer to run than min_loop_time, then it will
+    be run less often than once per min_loop_time.
+    """
+    def decorator(function: Callable) -> Callable:
+        @wraps(function)
+        def wrapper(*args, **kwargs):
+            while(True):
+                start_time = timezone.now()
+
+                function(*args, **kwargs)
+
+                loop_time = timezone.now() - start_time
+                if loop_time < min_loop_time:
+                    remaining_time = MIN_LOOP_TIME - loop_time
+                    time.sleep(remaining_time.seconds)
+
+        return wrapper
+    return decorator
+
+
+@do_forever(MIN_LOOP_TIME)
+def retry_failed_downloader_jobs() -> None:
+    """Handle downloader jobs that were marked as a failure."""
+    failed_jobs = DownloaderJob.objects.filter(success=False, retried=False)
+    handle_downloader_jobs(failed_jobs)
+
+
+@do_forever(MIN_LOOP_TIME)
+def retry_hung_downloader_jobs() -> None:
+    """Retry downloader jobs that were started but never finished."""
+    minimum_start_time = timezone.now() - MAX_DOWNLOADER_RUN_TIME
+    hung_jobs = DownloaderJob.objects.filter(
+        success=None,
+        retried=False,
+        end_time=None,
+        start_time__lt=minimum_start_time
+    )
+
+    handle_downloader_jobs(hung_jobs)
+
+
+@do_forever(MIN_LOOP_TIME)
+def retry_lost_downloader_jobs() -> None:
+    """Retry downloader jobs that went too long without being started.
+
+    Idea: at some point this function could integrate with the spot
+    instances to determine if jobs are hanging due to a lack of
+    instances. A naive time-based implementation like this could end
+    up retrying every single queued job if there were a long period of
+    spot instance bid price being very high.
```

Comment:
Thank you for the explanation. What I was confused about was the term of `spot instance bid price`. If I understand correctly, the last sentence actually means:
```
... if there were a long period during which the price of spot instance is higher than the bid price.
```
right?